File: algo-expert/min-heap.py
class MinHeap:
    def __init__(self, lst):
        self.heap = lst
        self.heapify()

    def heapify(self):
        for i in reversed(range(self.length())):
            self.siftUp(i)

    def siftUp(self, start_idx):
        swap_idx, node = start_idx, self.heap[start_idx]
        left_idx = 2*start_idx + 1

        while left_idx < self.length():
            right_idx = left_idx + 1
            if right_idx < self.length() and self.heap[left_idx] >= self.heap[right_idx]:
                left_idx = right_idx

            self.heap[swap_idx] = self.heap[left_idx]
            swap_idx = left_idx
            left_idx = 2*left_idx + 1

        # swap_idx is now a leaf node - place the node we're working with there
        # and sift it's parent nodes down if need be
        self.heap[swap_idx] = node
        self.siftDown(start_idx, swap_idx)

    def siftDown(self, start_idx, swap_idx):
        node = self.heap[swap_idx]
        while swap_idx > start_idx:
            parent_idx = (swap_idx-1)//2
            if node < self.heap[parent_idx]:
                self.heap[swap_idx] = self.heap[parent_idx]
                swap_idx = parent_idx
            else:
                break
        self.heap[swap_idx] = node

    def length(self):
        return len(self.heap)

    def insert(self, value):
        self.heap.append(value)
        self.siftDown(0, self.length() - 1)

    def pop(self):
        last_item = self.heap.pop()
        if self.heap:
            first_item = self.heap[0]
            self.heap[0] = last_item
            self.siftUp(0)
            return first_item
        return last_item


def heapSort(array):
    min_heap = MinHeap(array)
    result = []
    while min_heap:
        result.append(min_heap.pop())
    return result

mh = MinHeap([8, 5, 2, 9, 5, 6, 3])
print(mh.heap)

File: csharp-algo/csharp-algo/Collections/MinHeap.cs
using System;
using System.Collections.Generic;

namespace csharp_algo.Collections
{
    public class MinHeap<T> where T : IComparable<T>
    {
        public int Count => _heap.Count;
        private List<T> _heap;


        public MinHeap()
        {
            _heap = new List<T>();
        }

        public MinHeap(List<T> arr)
        {
            _heap = arr;
            Heapify();
        }

        public void Heapify()
        {
            for (var i = _heap.Count / 2; i >= 0; i--)
                SiftUp(i);
        }

        private void SiftDown(int startIdx, int swapIdx)
        {
            T node = _heap[swapIdx];
            while (startIdx < swapIdx)
            {
                int parentIdx = (swapIdx - 1) / 2;
                if (node.CompareTo(_heap[parentIdx]) == -1)
                {
                    _heap[swapIdx] = _heap[parentIdx];
                    swapIdx = parentIdx;
                    continue;
                }

                break;
            }

            _heap[swapIdx] = node;
        }

        private void SiftUp(int startIdx)
        {
            T node = _heap[startIdx];
            int swapIdx = startIdx;
            int leftIdx = startIdx * 2 + 1;
            while (leftIdx < _heap.Count)
            {
                int rightIdx = leftIdx + 1;
                if (rightIdx < _heap.Count && _heap[rightIdx].CompareTo(_heap[leftIdx]) <= 0)
                    leftIdx = rightIdx;
                _heap[swapIdx] = _heap[leftIdx];
                swapIdx = leftIdx;
                leftIdx = leftIdx * 2 + 1;
            }

            _heap[swapIdx] = node;
            SiftDown(startIdx, swapIdx);
        }

        public void Push(T val)
        {
            _heap.Add(val);
            SiftDown(0, _heap.Count - 1);
        }

        public T Pop()
        {
            T last = _heap[^1];
            _heap.RemoveAt(_heap.Count-1);
            if (_heap.Count == 0) return last;

            T first = _heap[0];
            _heap[0] = last;
            SiftUp(0);
            return first;
        }

        public T Peek()
        {
            return _heap[0];
        }
    }
}